File: PCFG.hs
module PCFG where

import Data.Map as M

type Nonterminal = String
type Terminal = String
type Prob = Double
data Rule = BranchRule Nonterminal Nonterminal Nonterminal | EndRule Nonterminal Terminal deriving (Eq,Ord,Show)
type PCFG = ([Nonterminal], [Terminal], Nonterminal -> Prob, Rule -> Prob)

-- Here's the PCFG from (2.57) on the handout. 
g0 :: PCFG
g0 = (  ["VP","NP","PP","V","P"], 
        ["watches","spies","telescopes","with"], 
        \n -> M.findWithDefault 0 n (M.fromList [("VP",1.0)]), 
        \r -> M.findWithDefault 0 r (M.fromList [(BranchRule "VP" "V" "NP",     0.4), 
                                                 (BranchRule "VP" "VP" "PP",    0.2), 
                                                 (EndRule    "VP" "watches",    0.3), 
                                                 (EndRule    "VP" "spies",      0.1), 
                                                 (BranchRule "NP" "NP" "PP",    0.2), 
                                                 (EndRule    "NP" "watches",    0.3), 
                                                 (EndRule    "NP" "spies",      0.2), 
                                                 (EndRule    "NP" "telescopes", 0.3), 
                                                 (BranchRule "PP" "P" "NP",     1.0), 
                                                 (EndRule    "V"  "watches",    1.0), 
                                                 (EndRule    "P"  "with",       1.0)
                                                ])
     )

--------------------------------------------------------------------------

-- This function calculates inside values following equation (2.69).
-- It works, but gets very slow if you try strings longer than about four or five symbols, because 
-- previously-computed values don't get re-used in the way they do if we work with a chart like in (2.68).
-- Example:
--  *PCFG> inside g0 ["watches","spies","with","telescopes"] "VP"
--  9.600000000000003e-3
inside :: PCFG -> [Terminal] -> Nonterminal -> Prob
inside g xs n =
    let (nonterms, sigma, initprob, ruleprob) = g in
    case xs of
    [] -> 0
    [x] -> ruleprob (EndRule n x)
    _ -> sum [ruleprob (BranchRule n l r) * inside g (take i xs) l * inside g (drop i xs) r | 
                    i <- [1 .. length xs - 1], 
                    l <- nonterms, 
                    r <- nonterms
             ]

-- This function calculates outside values following equation (2.74).
-- Example:
--  *PCFG> outside g0 (["watches"],["with","telescopes"]) "NP"
--  4.8e-2
--  *PCFG> outside g0 (["watches","spies"],[]) "PP"
--  3.200000000000001e-2
outside :: PCFG -> ([Terminal],[Terminal]) -> Nonterminal -> Prob
outside g (ys,zs) n =
    let (nonterms, sigma, initprob, ruleprob) = g in
    if ys == [] && zs == [] then
        initprob n
    else
        sum [ruleprob (BranchRule p n r) * inside g (take i zs) r * outside g (ys, drop i zs) p | 
                    i <- [1 .. length zs], 
                    p <- nonterms, 
                    r <- nonterms
            ]
        +
        sum [ruleprob (BranchRule p l n) * inside g (drop i ys) l * outside g (take i ys, zs) p | 
                    i <- [0 .. length ys - 1], 
                    p <- nonterms, 
                    l <- nonterms
            ]

--------------------------------------------------------------------------

-- This function constructs a chart like the one shown in (2.68), given a PCFG and a 
-- list of terminal symbols. 
-- The result is a Map that maps (symbol-list, nonterminal) pairs to probabilities. 
-- Example:
--  *PCFG> let c = insideChart g0 ["watches","spies","with","telescopes"] in (c ! (["watches","spies","with","telescopes"],"VP"))
--  9.600000000000003e-3
--  *PCFG> let c = insideChart g0 ["watches","spies","with","telescopes"] in (c ! (["spies","with","telescopes"],"NP"))
--  1.2000000000000002e-2

insideChart :: PCFG -> [Terminal] -> M.Map ([Terminal],Nonterminal) Prob
insideChart g string =

    let (nonterms, sigma, initprob, ruleprob) = g in

    -- Set up the list of cells to be filled, in order; 
    -- each cell is represented by a substring
    let cellsToFill = [take len (drop startpos string) | 
                            len <- [1 .. length string], 
                            startpos <- [0 .. length string - len]
                      ]
    in

    -- Define a function that will add the inside value for a single (string,nonterminal) pair 
    -- to an existing chart
    let oneInsideValue xs n chart =
            let prob = case xs of
                       [] -> 0
                       [x] -> ruleprob (EndRule n x)
                       _ -> sum [ruleprob (BranchRule n l r) * (M.findWithDefault 0 (take i xs, l) chart) * (M.findWithDefault 0 (drop i xs, r) chart) |
                                    i <- [1 .. length xs - 1], 
                                    l <- nonterms, 
                                    r <- nonterms
                                ]
            in
            if prob /= 0 then M.insert (xs,n) prob chart else chart
    in

    -- Define a function which takes a list of cells and a chart, and 
    -- fills in those cells in the given order
    let fillCells cells chart =
            case cells of
            [] -> chart
            xs:rest -> let newChart = Prelude.foldl (\ch -> \n -> oneInsideValue xs n ch) chart nonterms in
                       fillCells rest newChart
    in

    -- Now, fill in all the cells, starting with an empty chart
    fillCells cellsToFill M.empty

-- Total probability of a string, using inside values from a chart, following equation (2.66).
-- Example:
--  *PCFG> stringProbViaInsideChart g0 ["watches","spies","with","telescopes"]
--  9.600000000000003e-3
stringProbViaInsideChart :: PCFG -> [Terminal] -> Prob
stringProbViaInsideChart g xs =
    let (nonterms, sigma, initprob, ruleprob) = g in
    let chart = insideChart g xs in
    sum [initprob n * M.findWithDefault 0 (xs,n) chart | n <- nonterms]

--------------------------------------------------------------------------

-- This function constructs a chart of outside values, given a PCFG and a sequence of 
-- terminal symbols. We didn't discuss this in class, but it's analogous to constructing 
-- a chart of inside values, just in reverse; see p.353 and Figure 3.2 of Jelinek et al (1992). 
-- The result is a Map that maps ((symbol-list,symbol-list), nonterminal) pairs to probabilities. 
-- Example:
--  *PCFG> let c = outsideChart g0 ["watches","spies","with","telescopes"] in (c ! ((["watches"],["with","telescopes"]),"NP"))
--  4.8e-2
--  *PCFG> let c = outsideChart g0 ["watches","spies","with","telescopes"] in (c ! ((["watches","spies"],[]),"PP"))
--  3.200000000000001e-2

outsideChart :: PCFG -> [Terminal] -> M.Map (([Terminal],[Terminal]),Nonterminal) Prob
outsideChart g string =

    let (nonterms, sigma, initprob, ruleprob) = g in

    -- Get a completed chart of inside values
    let ichart = insideChart g string in

    -- Set up the list of cells to be filled, in order; 
    -- each cell is represented by a pair of strings (ys,zs)
    let cellsToFill = [((take startpos string, drop (startpos+len) string), n) |
                            len <- [length string, length string - 1 .. 1], 
                            startpos <- [0 .. length string - len], 
                            n <- nonterms
                      ]
    in

    -- Define a function that will add the outside value for a single ((string,string),nonterminal) pair 
    -- to an existing chart
    let oneOutsideValue (ys,zs) n chart =
            let prob =
                    if ys == [] && zs == [] then
                        initprob n
                    else
                        sum [ruleprob (BranchRule p n r) * (M.findWithDefault 0 (take i zs, r) ichart) * (M.findWithDefault 0 ((ys, drop i zs), p) chart) |
                                    i <- [1 .. length zs], 
                                    p <- nonterms, 
                                    r <- nonterms
                            ]
                        +
                        sum [ruleprob (BranchRule p l n) * (M.findWithDefault 0 (drop i ys, l) ichart) * (M.findWithDefault 0 ((take i ys, zs), p) chart) |
                                    i <- [0 .. length ys - 1], 
                                    p <- nonterms, 
                                    l <- nonterms
                            ]
            in
            if prob /= 0 then M.insert ((ys,zs),n) prob chart else chart
    in

    -- Define a function which takes a list of cells and a chart, and 
    -- fills in those cells in the given order
    let fillCells cells chart =
            case cells of
            [] -> chart
            ((ys,zs),n):rest -> let newChart = Prelude.foldl (\ch -> \n -> oneOutsideValue (ys,zs) n ch) chart nonterms in
                                fillCells rest newChart
    in

    -- Now, fill in all the cells, starting with an empty chart
    fillCells cellsToFill M.empty

File: pcfg.py

from collections import defaultdict

###############################################################################################

# Here's the PCFG from (2.57) on the handout. 
# I encode the functions I and R by wrapping a python dictionary in a little function that 
# looks up its argument in the dictionary and returns zero if no entry for it is found.

g0 = (  ["VP","NP","PP","V","P"] , 
        ["watches","spies","telescopes","with"] , 
        lambda nt: {"VP": 1.0}.get(nt,0) ,
        lambda *x: {("VP", ("V","NP")):   0.4, 
                    ("VP", ("VP","PP")):  0.2, 
                    ("VP", "watches"):    0.3, 
                    ("VP", "spies"):      0.1, 
                    ("NP", ("NP","PP")):  0.2, 
                    ("NP", "watches"):    0.3, 
                    ("NP", "spies"):      0.2, 
                    ("NP", "telescopes"): 0.3, 
                    ("PP", ("P","NP")):   1.0, 
                    ("V",  "watches"):    1.0, 
                    ("P",  "with"):       1.0, 
                   }.get(x,0)
     )

###############################################################################################

# This function calculates inside values (given a grammar g, a string xs and a nonterminal symbol n) 
# following equation (2.69) in a completely direct way.
# It works, but gets very slow if you try strings longer than about four or five symbols, because 
# previously-computed values don't get re-used the way they do if we work with a chart like in (2.68).
# Example:
#   >>> inside(g0, ["watches","spies","with","telescopes"], "VP")
#   0.009600000000000003
def inside(g, xs, n):
    (nonterms, sigma, initprob, ruleprob) = g
    if len(xs) == 1:
        return ruleprob(n, xs[0])
    else:
        return sum([ruleprob(n,(l,r)) * inside(g, xs[:i], l) * inside(g, xs[i:], r) 
                        for i in range(1,len(xs)) 
                        for l in nonterms 
                        for r in nonterms
                   ])

# This function calculates outside values (given a grammar g, a pair of strings 
# and a nonterminal symbol n) following equation (2.74) directly.
# Example:
#   >>> outside(g0, (["watches"],["with","telescopes"]), "NP")
#   0.048
#   >>> outside(g0, (["watches","spies"],[]), "PP")
#   0.03200000000000001
def outside(g, stringpair, n):
    (ys,zs) = stringpair
    (nonterms, sigma, initprob, ruleprob) = g
    if ys == [] and zs == []:
        return initprob(n)
    else:
        return (
            sum([ruleprob(p,(n,r)) * inside(g, zs[:i], r) * outside(g, (ys,zs[i:]), p) 
                    for i in range(1,len(zs)+1) 
                    for p in nonterms 
                    for r in nonterms
                ])
            +
            sum([ruleprob(p,(l,n)) * inside(g, ys[i:], l) * outside(g, (ys[:i],zs), p) 
                    for i in range(0,len(ys)) 
                    for p in nonterms 
                    for l in nonterms
                ])
        )

###############################################################################################

# This function constructs a chart like the one shown in (2.68), given a PCFG and a 
# sequence of terminal symbols. 
# The result is a dictionary that maps (symbol-sequence, nonterminal) pairs to probabilities. 
# The symbol-sequence when doing a lookup has to be a tuple, not a list, because python's lists are mutable and not hashable.
# Example:
#   >>> c = inside_chart(g0, ["watches","spies","with","telescopes"])
#   >>> c[(("watches","spies","with","telescopes"),"VP")]
#   0.009600000000000003
#   >>> c[(("spies","with","telescopes"),"NP")]
#   0.012000000000000002

def inside_chart(g, string):

    (nonterms, sigma, initprob, ruleprob) = g
    chart = {}

    # Each diagonal of the CKY chart represents substrings of a certain length.
    # We loop over these diagonals by looping over lengths from 1 up to the length of the entire string.
    for length in range(1, len(string)+1):

        # Each cell on this diagonal represents a particular substring of the chosen length.
        # We loop over these substrings by looping over starting positions for substrings, i.e. by looping 
        # over rows of the chart (but limited to those rows that contain a cell on this diagonal).
        for startpos in range(0, len(string)-length+1):

            # This is the substring/infix represented by this cell
            xs = tuple(string[startpos:startpos+length])

            for n in nonterms:

                # Calculate inside(xs)(n) using equation (2.69)
                if length == 1:
                    p = ruleprob(n, xs[0])
                else:
                    p = sum([ruleprob(n,(l,r)) * chart[(xs[:i],l)] * chart[(xs[i:],r)] 
                                for i in range(1,length) 
                                for l in nonterms 
                                for r in nonterms
                            ])

                # Store the result in the chart
                chart[(xs,n)] = p

    # Just for convenience: make all the zero entries in the chart implicit before returning it
    tidy_chart = defaultdict(float, [(key,p) for (key,p) in chart.items() if p != 0])
    return tidy_chart

# Total probability of a string, using inside values from a chart, following equation (2.66).
# Example:
#   >>> stringprob_via_inside_chart(g0, ["watches","spies","with","telescopes"])
#   0.009600000000000003
def stringprob_via_inside_chart(g, xs):
    (nonterms, sigma, initprob, ruleprob) = g
    chart = inside_chart(g, xs)
    return sum([initprob(n) * chart[(tuple(xs),n)] for n in nonterms])

###############################################################################################

# This function constructs a chart of outside values, given a PCFG and a sequence of 
# terminal symbols. We didn't discuss this in class, but it's analogous to constructing 
# a chart of inside values, just in reverse; see p.353 and Figure 3.2 of Jelinek et al (1992). 
# The result is a dictionary that maps ((symbol-sequence,symbol-sequence), nonterminal) pairs 
# to probabilities. 
# Example:
#   >>> oc = outside_chart(g0, ["watches","spies","with","telescopes"])
#   >>> oc[((("watches",),("with","telescopes")),"NP")]
#   0.048
#   >>> oc[((("watches","spies"),()),"PP")]
#   0.03200000000000001

def outside_chart(g, string):

    (nonterms, sigma, initprob, ruleprob) = g
    ochart = {}

    # First get a complete chart of all relevant inside values, which we will need
    ichart = inside_chart(g, string)

    # For each diagonal of the chart, now working down and left from the top-right corner
    for length in range(len(string), 0, -1):

        # For each row of the chart for which there is a cell on this diagonal
        for startpos in range(0, len(string)-length+1):

            ys = tuple(string[:startpos])
            zs = tuple(string[startpos+length:])

            for n in nonterms:

                # Calculate outside(ys,zs)(n) using equation (2.74)
                if len(ys) == 0 and len(zs) == 0:
                    prob = initprob(n)
                else:
                    prob = (sum([ruleprob(p,(n,r)) * ichart[(zs[:i],r)] * ochart[((ys,zs[i:]),p)]
                                    for i in range(1,len(zs)+1) 
                                    for p in nonterms 
                                    for r in nonterms
                                ])
                            +
                            sum([ruleprob(p,(l,n)) * ichart[(ys[i:],l)] * ochart[((ys[:i],zs),p)] 
                                    for i in range(0,len(ys)) 
                                    for p in nonterms 
                                    for l in nonterms
                                ])
                           )

                # Store the result in the chart
                ochart[((ys,zs),n)] = prob

    # Just for convenience: make all the zero entries in the chart implicit before returning it
    tidy_ochart = defaultdict(float, [(key,p) for (key,p) in ochart.items() if p != 0])
    return tidy_ochart

